docs/configuration.md
# Overview

The logdata-anomaly-miner can be configured in two different formats:
**yaml** and **python**. The preferred format is yaml and the default
configuration file for it is */etc/aminer/config.yaml*. The python
format can be configured in */etc/aminer/config.py* and offers advanced
possibilities to configure the logdata-anomaly-miner. However, this is
only recommended for experts, as no errors are caught in the python
configuration, which can make debugging very difficult. For both formats
there are template configurations in */etc/aminer/template_config.yaml*
and */etc/aminer/template_config.py*.

The basic structure of the logdata-anomaly-miner is illustrated in the
folloging diagram:

![Structure of the configuration-file: GENERAL, INPUT, PARSING,
ANALYSING, EVENTHANDLING](images/aminer-config-color.png)

## Analysis Pipeline

The core component of the logdata-anomaly-miner is the "analysis
pipeline". It consists of the parts INPUT, ANALYSIS and OUTPUT.

![Parts of the analysis-pipeline](images/analysis-pipeline.png)

# Command-line Parameters

## -h, --help

Show the help message and exit.

## -v, --version

Show program's version number and exit.

## -u, --check-updates

Check if updates for the aminer are available and exit.

## -c CONFIG, --config CONFIG

  - Default: /etc/aminer/config.yml

Use the settings of the file CONFIG on startup. Two config-variants are
allowed: python and yaml.

!!! seealso "See also"

    [Overview](#overview)

## -D, --daemon

Run aminer as a daemon process.

## -s {0,1,2}, --stat {0,1,2}

Set the stat level. Possible stat-levels are 0 for no statistics, 1 for
normal statistic level and 2 for verbose statistics.

## -d {0,1,2}, --debug {0,1,2}

Set the debug level. Possible debug-levels are 0 for no debugging, 1 for
normal output (INFO and above), 2 for printing all debug information.

## --run-analysis

Run aminer analysis-child.

!!! note

    This parameter is for internal use only.

## -C, --clear

Remove all persistence directories and run aminer.

## -r REMOVE, --remove REMOVE

Remove a specific persistence directory. REMOVE must be the name of the
directory and must not contain '/' or '.'. Usually this directory can be
found in '/var/lib/aminer'.

## -R RESTORE, --restore RESTORE

Restore a persistence backup. RESTORE must be the name of the directory
and must not contain '/' or '.'. Usually this directory can be found in
'/var/lib/aminer'.

## -f, --from-begin

Removes repositioning data before starting the aminer so that all input
files will be analyzed starting from the first line in the file rather
than the last previously analyzed line.

## -o, --offline-mode

Stop the aminer after all logs have been processed.

!!! note

    This parameter is useful for forensic analysis.

## --config-properties KEY=VALUE [KEY=VALUE ...]

Set a number of config_properties by using key-value pairs (do not put
spaces before or after the = sign). If a value contains spaces, you
should define it with double quotes: 'foo="this is a sentence". Note
that values are always treated as strings. If values are already defined
in the config_properties, the input types are converted to the ones
already existing.

# Configuration Reference

## General Configuration

### LearnMode

  - Type: boolean (True,False)
  - Default: False

This options turns the LearnMode on globally.

!!! warning

    This option can be overruled by the learn_mode that is configurable per
    analysis component.

``` yaml
LearnMode: True
```

### AminerUser

  - Default: aminer

This option defines the system-user that owns the aminer-process.

``` yaml
AminerUser: 'aminer'
```

### AminerGroup

  - Default: aminer

This option defines the system-group that owns the aminer-process.

``` yaml
AminerGroup: 'aminer'
```

### AnalysisConfigFile

  - Default: None

This (optional) configuration file contains the whole analysis child
configuration (code). When missing those configuration parameters are
also taken from the main config.

!!! warning

    This option is only available for python configs. It does not work for
    yaml configs.

``` python
config_properties['AnalysisConfigFile'] = 'analysis.py'
```

### RemoteControlSocket

This option controls where the unix-domain-socket for the RemoteControl
should be created. The socket will not be created if this option is not
set.

``` yaml
RemoteControlSocket: '/var/lib/aminer/remcontrol.sock'
```

### SuppressNewMatchPathDetector

  - Default: False
  - Type: boolean (True,False)

Disable the output of the NewMatchPathDetector which detects new paths
for logtypes.

``` yaml
SuppressNewMatchPathDetector: False
```

### LogResourceList

  - Required: **True**
  - Resource-Types: `file://`, `unix://`

Define the list of log resources to read from: the resources named here
do not need to exist when aminer is started. This will just result in a
warning. However if they exist, they have to be readable by the aminer
process\! Every resource needs to define the `url` with the
resource-type. Optionally every resource can define `json` parameter
(boolean) to define if the resource input data is json and `parser_id`
to define the parser which should process the log data from this
resource. By default the `json_format` parameter in the `input` section
is used to determine if the input data is json or not.

Supported types are:

  - [file://\[path\]](file://%5Bpath%5D): Read data from file, reopen it
    after rollover
  - unix://\[path\]: Open the path as UNIX local socket for reading

<!-- end list -->

``` yaml
LogResourceList:
    - url: 'file:///var/log/apache2/access.log'
    - url: 'file:///home/ubuntu/data/mail.cup.com-train/daemon.log'
      json: True
      parser_id: 'syslog_parser'
    - url: 'file:///home/ubuntu/data/mail.cup.com-train/auth.log'
    - url: 'file:///home/ubuntu/data/mail.cup.com-train/suricata/eve.json'
    - url: 'file:///home/ubuntu/data/mail.cup.com-train/suricata/fast.log'
      json: True
      parser_id: 'suricata_fastlog'
```

### Core.PersistenceDir

  - Default: /var/lib/aminer

Read and store information to be used between multiple executions of
aminer in this directory. The directory must only be accessible to the
'AminerUser' but not group/world readable. On violation, aminer will
refuse to start.

``` yaml
Core.PersistenceDir: '/var/lib/aminer'
```

### Core.PersistencePeriod

  - Type: Number of seconds
  - Default: 600

This options controls whether the logdata-anomaly-miner should write its
persistency to disk.

``` yaml
Core.PersistencePeriod: 600
```

### Core.LogDir

  - Default: /var/lib/aminer/log

Directory for logfiles. This directory must be writeable to the
'AminerUser'.

``` yaml
Core.LogDir: '/var/lib/aminer/log'
```

### MailAlerting.TargetAddress

  - Default: disabled

Define a target e-mail address to send alerts to. When undefined, no
e-mail notification hooks are added.

``` yaml
MailAlerting.TargetAddress: 'root@localhost'
```

### MailAlerting.FromAddress

Sender address of e-mail alerts. When undefined, "sendmail"
implementation on host will decide, which sender address should be used.

``` yaml
MailAlerting.FromAddress: 'root@localhost'
```

### MailAlerting.SubjectPrefix

  - Default: "aminer Alerts"

Define, which text should be prepended to the standard aminer subject.

``` yaml
MailAlerting.SubjectPrefix: 'aminer Alerts:'
```

### MailAlerting.AlertGraceTime

  - Type: Number of seconds
  - Default: 0 (any event can immediately trigger alerting)

Define a grace time after startup before aminer will react to an event
and send the first alert e-mail.

``` yaml
MailAlerting.AlertGraceTime: 0
```

### MailAlerting.EventCollectTime

  - Type: Number of seconds
  - Default: 10

Define how many seconds to wait after a first event triggered the
alerting procedure before really sending out the e-mail. In that
timespan, events are collected and will be sent all using a single
e-mail.

``` yaml
MailAlerting.EventCollectTime: 10
```

### MailAlerting.MinAlertGap

  - Type: Number of seconds
  - Default: 600

Define the minimum time between two alert e-mails in seconds to avoid
spamming. All events during this timespan are collected and sent out
with the next report.

``` yaml
MailAlerting.MinAlertGap: 600
```

### MailAlerting.MaxAlertGap

  - Type: Number of seconds
  - Default: 600

Define the maximum time between two alert e-mails in seconds. When
undefined this defaults to "MailAlerting.MinAlertGap". Otherwise this
will activate an exponential backoff to reduce messages during permanent
error states by increasing the alert gap by 50% when more alert-worthy
events were recorded while the previous gap time was not yet elapsed.

``` yaml
MailAlerting.MaxAlertGap: 600
```

### MailAlerting.MaxEventsPerMessage

  - Type: Number of events
  - Default: 1000

Define how many events should be included in one alert mail at most.

``` yaml
MailAlerting.MaxEventsPerMessage: 1000
```

### LogPrefix

This option defines the prefix for the output of each anomaly.

``` yaml
LogPrefix: ''
```

### Log.Encoding

  - Type: string
  - Default: 'utf-8'

This option defines the encoding of the logfiles.

``` yaml
Log.Encoding: 'utf-8'
```

### Log.StatisticsPeriod

  - Type: Number of seconds
  - Default: 3600

Defines how often to write into stat-logfiles.

``` yaml
Log.StatisticsPeriod: 3600
```

### Log.StatisticsLevel

  - Type: Number of loglevel
  - Default: 1

Defines the loglevel for the stat logs.

``` yaml
Log.StatisticsLevel: 2
```

### Log.DebugLevel

  - Type: Number of loglevel
  - Default: 1

Defines the loglevel of the aminer debug-logfile.

``` yaml
Log.DebugLevel: 2
```

### Log.RemoteControlLogFile

  - Type: string (path to the logfile)
  - Default: '/var/lib/aminer/log/aminerRemoteLog.log'

Defines the path of the logfile for the RemoteControl.

``` yaml
Log.RemoteControlLogFile: '/var/log/aminerremotecontrol.log'
```

### Log.StatisticsFile

  - Type: string (path to the logfile)
  - Default: '/var/lib/aminer/log/statistics.log'

Defines the path of the stats-file.

``` yaml
Log.StatisticsFile: '/var/log/aminer-stats.log'
```

### Log.DebugFile

  - Type: string (path to the logfile)
  - Default: '/var/lib/aminer/log/aminer.log'

Defines the path of the debug-log-file.

``` yaml
Log.DebugFile: '/var/log/aminer.log'
```

### Log.Rotation.MaxBytes

  - Type: number of bytes
  - Default: 1048576 (1 Megabyte)

Defines the number of bytes before "Log.RemoteControlLogFile",
"Log.StatisticsFile" and "Log.DebugFile" is rotated.

``` yaml
Log.Rotation.MaxBytes: 1048576
```

### Log.Rotation.BackupCount

  - Type: number of old logfiles
  - Default: 5

Defines the number of logfiles saved after rotation of
"Log.RemoteControlLogFile", "Log.StatisticsFile" and "Log.DebugFile".

``` yaml
Log.Rotation.BackupCount: 5
```

## Input

### timestamp_paths

  - Type: string or list of strings

Parser paths to DateTimeModelElements to set timestamp of log events.

``` yaml
timestamp_paths: '/model/time'
```

``` yaml
timestamp_paths:
   - '/parser/model/time'
   - '/parser/model/type/execve/time'
   - '/parser/model/type/proctitle/time'
   - '/parser/model/type/syscall/time'
   - '/parser/model/type/path/time'
```

### multi_source

  - Type: boolean (True,False)
  - Default: False

Flag to enable chronologically correct parsing from multiple
input-logfiles.

``` yaml
multi_source: True
```

### eol_sep

  - Default: 'n'

End of Line seperator for events.

!!! note

    Enables parsing of multiline logs.

``` yaml
eol_sep: '\r\n'
```

### json_format

  - Type: boolean (True,False)
  - Default: False

Enables parsing of logs in json-format.

``` yaml
json_format: True
```

### suppress_unparsed

  - Default: False

Boolean value that allows to suppress anomaly output about unparsed log
atoms.

``` yaml
suppress_unparsed: True
```

## Parsing

There are some predefined standard-model-elements like
*IpAddressDataModelElement*, *DateTimeModelElement*,
*FixedDataModelElement* and so on. They are located in the
python-source-tree of logdata-anomaly-miner. A comprehensive list of all
possible standard-model-elements can be found below. Using these
standard-model-elements it is possible to create custom parser models.
Currently there are two methods of doing it:

1.  Using a python-script that is located in */etc/aminer/conf-enabled*:

<!-- end list -->

``` python
""" /etc/aminer/conf-enabled/ApacheAccessParsingModel.py"""
from aminer.parsing.DateTimeModelElement import DateTimeModelElement
from aminer.parsing.DecimalIntegerValueModelElement import DecimalIntegerValueModelElement
from aminer.parsing.DelimitedDataModelElement import DelimitedDataModelElement
from aminer.parsing.FirstMatchModelElement import FirstMatchModelElement
from aminer.parsing.FixedDataModelElement import FixedDataModelElement
from aminer.parsing.FixedWordlistDataModelElement import FixedWordlistDataModelElement
from aminer.parsing.IpAddressDataModelElement import IpAddressDataModelElement
from aminer.parsing.OptionalMatchModelElement import OptionalMatchModelElement
from aminer.parsing.SequenceModelElement import SequenceModelElement
from aminer.parsing.VariableByteDataModelElement import VariableByteDataModelElement

def get_model():
    """Return a model to parse Apache Access logs from the AIT-LDS."""
    alphabet = b'!"#$%&\'()*+,-./0123456789:;<>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ\\^_`abcdefghijklmnopqrstuvwxyz{|}~=[]'

    model = SequenceModelElement('model', [
        FirstMatchModelElement('client_ip', [
            IpAddressDataModelElement('client_ip'),
            FixedDataModelElement('localhost', b'::1')
            ]),
        FixedDataModelElement('sp1', b' '),
        VariableByteDataModelElement('client_id', alphabet),
        FixedDataModelElement('sp2', b' '),
        VariableByteDataModelElement('user_id', alphabet),
        FixedDataModelElement('sp3', b' ['),
        DateTimeModelElement('time', b'%d/%b/%Y:%H:%M:%S'),
        FixedDataModelElement('sp4', b' +'),
        DecimalIntegerValueModelElement('tz'),
        FixedDataModelElement('sp5', b'] "'),
        FirstMatchModelElement('fm', [
            FixedDataModelElement('dash', b'-'),
            SequenceModelElement('request', [
                FixedWordlistDataModelElement('method', [
                    b'GET', b'POST', b'PUT', b'HEAD', b'DELETE', b'CONNECT', b'OPTIONS', b'TRACE', b'PATCH']),
                FixedDataModelElement('sp6', b' '),
                DelimitedDataModelElement('request', b' ', b'\\'),
                FixedDataModelElement('sp7', b' '),
                DelimitedDataModelElement('version', b'"'),
                ])
            ]),
        FixedDataModelElement('sp8', b'" '),
        DecimalIntegerValueModelElement('status_code'),
        FixedDataModelElement('sp9', b' '),
        DecimalIntegerValueModelElement('content_size'),
        OptionalMatchModelElement(
            'combined', SequenceModelElement('combined', [
                FixedDataModelElement('sp10', b' "'),
                DelimitedDataModelElement('referer', b'"', b'\\'),
                FixedDataModelElement('sp11', b'" "'),
                DelimitedDataModelElement('user_agent', b'"', b'\\'),
                FixedDataModelElement('sp12', b'"'),
                ])),
        ])

    return model
```

This parser can be used as "type" in **/etc/aminer/config.yml**:

``` yaml
Parser:
     - id: 'apacheModel'
       type: ApacheAccessModel
       name: 'apache'
```

!!! warning

    Please do not create files with the ending "ModelElement.py" in
    /etc/aminer/conf-enabled\!

2.  Configuring the parser-model inline in **/etc/aminer/config.yml**

<!-- end list -->

``` yaml
Parser:
        - id: host_name_model
          type: VariableByteDataModelElement
          name: 'host'
          args: '-.01234567890abcdefghijklmnopqrstuvwxyz:'

        - id: identity_model
          type: VariableByteDataModelElement
          name: 'ident'
          args: '-.01234567890abcdefghijklmnopqrstuvwxyz:'

        - id: user_name_model
          type: VariableByteDataModelElement
          name: 'user'
          args: '0123456789abcdefghijklmnopqrstuvwxyz.-'

        - id: new_time_model
          type: DateTimeModelElement
          name: 'time'
          date_format: '[%d/%b/%Y:%H:%M:%S +0000]'

        - id: sq3
          type: FixedDataModelElement
          name: 'sq3'
          args: ' "'

        - id: request_method_model
          type: FixedWordlistDataModelElement
          name: 'method'
          args:
                  - 'GET'
                  - 'POST'
                  - 'PUT'
                  - 'HEAD'
                  - 'DELETE'
                  - 'CONNECT'
                  - 'OPTIONS'
                  - 'TRACE'
                  - 'PATCH'

        - id: request_model
          type: VariableByteDataModelElement
          name: 'request'
          args: '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ.-/()[]{}!$%&=<?*+'

        - id: http1
          type: FixedDataModelElement
          name: 'http1'
          args: ' HTTP/'

        - id: version_model
          type: VariableByteDataModelElement
          name: 'version'
          args: '0123456789.'

        - id: sq4
          type: FixedDataModelElement
          name: 'sq4'
          args: '" '

        - id: status_code_model
          type: DecimalIntegerValueModelElement
          name: 'status'

        - id: size_model
          type: DecimalIntegerValueModelElement
          name: 'size'

        - id: sq5
          type: FixedDataModelElement
          name: 'sq5'
          args: ' "-" "'

        - id: user_agent_model
          type: VariableByteDataModelElement
          name: 'useragent'
          args: '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ.-/()[]{}!$%&=<?*+;:_ '

        - id: sq6
          type: FixedDataModelElement
          name: 'sq6'
          args: '"'

        - id: 'startModel'
          start: True
          type: SequenceModelElement
          name: 'accesslog'
          args:
                  - host_name_model
                  - WHITESPACE
                  - identity_model
                  - WHITESPACE
                  - user_name_model
                  - WHITESPACE
                  - new_time_model
                  - sq3
                  - request_method_model
                  - WHITESPACE
                  - request_model
                  - http1
                  - version_model
                  - sq4
                  - status_code_model
                  - WHITESPACE
                  - size_model
                  - sq5
                  - user_agent_model
                  - sq6
```

The parsing section in **/etc/aminer/config.yml** starts with the
statement "Parser:" followed by a list of parser-models. Every
parser-model in this list must have a unique **id** and a **type**. The
unique **id** can be used to cascade models by adding the **id** of an
parser-model as arguments(**args**). One parser of this list must
contain <span class="title-ref">start: True</span> that indicates the
root of the parser tree:

``` yaml
Parser:
     - id: 'apacheModel'
       type: ApacheAccessModel
       name: 'apache'

     - id: 'startModel'
       start: True
       type: SequenceModelElement
       name: 'model'
       args: apacheModel
```

  - **id**: must be a unique string
  - **type**: must be an existing ModelElement
  - **name**: string with the element name
  - **start**: a boolean value that indicates the starting model. Only
    one parser-model must have enabled this option\!
  - **args**\*: a string or a list of strings containing the arguments
    of the specific parser.

!!! note

    args can contain the constant WHITESPACE which is a preset for spaces

### AnyByteDataModelElement

This parsing-element matches any byte but at least one. Thus a match
will always span the complete data from beginning to end.

``` yaml
Parser:
     - id: 'anyModel'
       type: AnyByteDataModelElement
       name: 'anymodel'
```

### Base64StringModelElement

This parsing-element matches base64 strings.

``` yaml
Parser:
     - id: 'anyModel'
       type: Base64StringModelElement
       name: 'b64model'
```

### DateTimeModelElement

This element parses dates using a custom, timezone and locale-aware
implementation similar to strptime.

  - **args**: a string or list containing the following parameters:
    1.    - date_format:  
            Is a string that represents the date format for parsing, see
            Python strptime specification for available formats.
            Supported format specifiers are:
            
            >   - %b: month name in current locale
            >   - %d: day in month, can be space or zero padded when
            >     followed by separator or at end of string.
            >   - %f: fraction of seconds (the digits after the the '.')
            >   - %H: hours from 00 to 23
            >   - %M: minutes
            >   - %m: two digit month number
            >   - %S: seconds
            >   - %s: seconds since the epoch (1970-01-01)
            >   - %Y: 4 digit year number
            >   - %z: detect and parse timezone strings like UTC, CET,
            >     +0001, etc. automatically.
            
              - Common formats are:
                
                  - '%b %d %H:%M:%S' e.g. for 'Nov 19 05:08:43'
                  - '%d.%m.%YT%H:%M:%S' e.g. for '07.02.2019T11:40:00'
                  - '%d.%m.%Y %H:%M:%S.%f' e.g. for '07.02.2019
                    11:40:00.123456'
                  - '%d.%m.%Y %H:%M:%S%z' e.g. for '07.02.2019
                    11:40:00+0000" or "07.02.2019 11:40:00 UTC'
                  - '%d.%m.%Y' e.g. for '07.02.2019'
                  - '%H:%M:%S' e.g. for '11:40:23'
    
    2.    - time_zone:  
            time_zone the timezone for parsing the values. Default:
            **UTC**. Within the yaml configuration, only values from
            **pytz.all_timezones** are accepted as time_zone value.
    
    3.    - text_local:  
            the locale to use for parsing the day and month names.
            Default: **system-locale**
    
    4.    - start_year:  
            start_year when parsing date records without any year
            information, assume this is the year of the first value
            parsed.

changelog
logdata-anomaly-miner (2.8.0) unstable; urgency=low
  Bugfixes:
  * fix suspendModeTest
  * fix issues with np.math
  * fix AMINERSRC variable in aminer_install script
  * fix error message in ByteStreamLineAtomizer
  * Fix stop_learning_time
  * Fix closing of Zmq producer socket at graceful shutdown
  * Fix sorting of None values in NMPVCD.do_persist()

  Changes: 
  * Rewrite of aminerwrapper
  * Set latest working python-kafka version
  * Remove restrictive flags in aminer, so the conf-enabled directory is not needed any more -> still need to keep conf-enabled to use custom parsing models
  * Allow resource_name to be of type bytes
  * Allow ParserMatch to be None in LogAtom
  * Improve LogResourceList error message when url does not start with file:// or unix://
  * Add an optional log line identifier to json output
  * Remove proposed AminerStartTimestamp parameter and implement dynamic per detector stop_learning_time setup in NMPD
  * Implement dynamic per detector stop_learning_time setup
  * Add comment explaining the lambda function in the NMPVCD.do_persist() method


 -- Markus Wurzenberger <markus.wurzenberger@ait.ac.at> Thu,  01 May 2025 12:00:00 +0000




logdata-anomaly-miner (2.7.0) unstable; urgency=low
  Bugfixes:
  * Fix broken links to python-modules
  * Fix initalization time index in EFD
  * Fix correct time index when learn mode is off
  * Fix edge case where confidence is divided by 0 in EventFrequencyDetector
  * Fix time until learn_mode is switched automatically
  * Fix issue where numerical values could not be used in CharsetDetector and EntropyDetector
  * Fix feature list order in PCA detector
  * Fix time index persistence in EFD
  * Fix last seen log in EFD
  * Fix newline parsing in json string values.
  * Fix issue of nested optional json dictionaries not parsing.

  Changes:
  * Implemented mechanism to allow granular configuration of LogResources.
  * Add seasonality to EFD
  * Add atom time in reset counter method
  * Add output_event_handlers parameter to the unparsed event handlers.
  * Add stop_learning_time and stop_learning_no_anomaly_time to schemas
  * Change id_path_list parameter to be optional in ValueRangeDetector.
  * Imporove JsonModelElement parser
  * Add learn mode and detection thresholds in MatchValueAverageChangeDetector
  * Implement AminerId config property
  * Add LogResource to the json output of anomalies
  * Add use_real_time parameter to yaml config
  * Move manpages to Debian independent directory
  * Add ignore_log_resource parameter to analysis components.
  * Improved performance of JsonModelElement
  * Add XmlModelElement
  * Implement stop_learning_time parameters in MatchValueAverageChangeDetector.
  * Add support for Debian Bookworm
  * Updated to urllib 1.26.19
  * Updated scipy to 1.10.0
  * Extend missing timestamp warning in ByteStreamLineAtomizer.
  * Enhance timezones in DateTimeModelElement
  * Increase default persistence time
  * Add support for Redhat based Linux
  * Improved documenation
  

 -- Markus Wurzenberger <markus.wurzenberger@ait.ac.at> Thu,  11 Jul 2024 12:00:00 +0000


logdata-anomaly-miner (2.6.1) unstable; urgency=low
  Changes:
  * minor refactoring
  * fixed code styling issues

 -- Markus Wurzenberger <markus.wurzenberger@ait.ac.at> Tue,  21 Feb 2023 12:00:00 +0000

logdata-anomaly-miner (2.6.0) unstable; urgency=low
  Bugfixes:
  * fixed bug in JsonModelElement where the aminer gets stuck in an endless loop searching for \x.
  * added input file path sanitization and fixed exception handling.
  * fixed a test for the remote control save config method.
  * fixed bug, that occured when starting one of the detectors VTD, VCD and TSA with an already existing persistency of the ETD, but not of the detectors.
  * fixed the MissingMatchPathValueDetector by comparing the detector_info[0] instead of the old_last_seen_timestamp.
  * ParserCount: Fixed timestamp in output
  * implemented the output_logline parameter in the NewMatchPathValueDetector.
  * fixed bug where the MissingMatchPathListValueDetector could not be used in yaml, because the ConfigValidator could not load the module.
  * runHowToEntropyDetector had missing permissions on CFG_PATH in some lines.
  * fixed bug with closing the streams.
  Changes:
  * renamed schemas to python files.
  * enabled systemd autorestart
  * improved documentation
  * added SlidingEventFrequencyDetector
  * added timestamp_scale parameter to the DateTimeModelElement.
  * added unique path param for EFD
  * added check so EXP_TYPE_MANDATORY is enforced.
  * replace raw data output with last log of event type rather than end of time window
  * added event count cluster detector
  * added experimental jsonstringparser
  * improved parameter consistency
  * added ScoringEventHandler
  * EFD: Added the functionality to analze the scoring_path_list with the ScoringEventHandler
  * ETD/TSA: Moved the initialization part of the TSA from the ETD to the TSA
  * support for ZeroMQ-Eventhandler
  * added support for named-pipes

 -- Markus Wurzenberger <markus.wurzenberger@ait.ac.at> Fri,  20 Jan 2023 12:00:00 +0000

logdata-anomaly-miner (2.5.1) unstable; urgency=low
  Bugfixes:
  * EFD: Fixed problem that appears with empty windows
  * Fixed index out of range if matches are empty in JsonModelElement array.
  * EFD: Fixed problem that appears with empty windows
  * EFD: Enabled immediate detection without training, if both limits are set
  * EFD: Fixed bug related to auto_include_flag
  * Remove spaces in aminer logo
  * ParserCounter: Fixed do_timer
  * Fixed code to allow the usage of AtomFilterMatchAction in yaml configs
  * Fixed JsonModelElement when json object is null
  * Fix incorrect message of charset detector
  * Fix match list handling for json objects
  * Fix incorrect message of charset detector
  Changes:
  * Added nullable functionality to JsonModelElements
  * Added include-directive to supervisord.conf

aecid-testsuite/unit/data/config.py
# This is a template for the "aminer" logfile miner tool. Copy
# it to "config.py" and define your ruleset.

config_properties = {}

# Define the list of log resources to read from: the resources
# named here do not need to exist when aminer is started. This
# will just result in a warning. However if they exist, they have
# to be readable by the aminer process! Supported types are:
# * file://[path]: Read data from file, reopen it after rollover
# * unix://[path]: Open the path as UNIX local socket for reading
config_properties['LogResourceList'] = ['file:///tmp/syslog']

# Define the uid/gid of the process that runs the calculation
# after opening the log files:
config_properties['AminerUser'] = 'aminer'
config_properties['AminerGroup'] = 'aminer'

# Define the path, where aminer will listen for incoming remote
# control connections. When missing, no remote control socket
# will be created.
# config_properties['RemoteControlSocket'] = '/var/run/aminer-remote.socket'

# Read the analyis from this file. That part of configuration
# is separated from the main configuration so that it can be loaded
# only within the analysis child. Non-absolute path names are
# interpreted relatively to the main configuration file (this
# file). When empty, this configuration has to contain the configuration
# for the child also.
# config_properties['AnalysisConfigFile'] = 'analysis.py'

# Read and store information to be used between multiple invocations
# of py in this directory. The directory must only be accessible
# to the 'AminerUser' but not group/world readable. On violation,
# py will refuse to start. When undefined, '/var/lib/aminer'
# is used.
config_properties['Core.PersistenceDir'] = '/tmp/lib/aminer'

# Define a target e-mail address to send alerts to. When undefined,
# no e-mail notification hooks are added.
config_properties['MailAlerting.TargetAddress'] = 'mail@localhost'
# Sender address of e-mail alerts. When undefined, "sendmail"
# implementation on host will decide, which sender address should
# be used.
config_properties['MailAlerting.FromAddress'] = 'mail@localhost'
# Define, which text should be prepended to the standard aminer
# subject. Defaults to "py Alerts:"
config_properties['MailAlerting.SubjectPrefix'] = 'aminer Alerts:'
# Define a grace time after startup before aminer will react to
# an event and send the first alert e-mail. Defaults to 0 (any
# event can immediately trigger alerting).
config_properties['MailAlerting.AlertGraceTime'] = 0
# Define how many seconds to wait after a first event triggered
# the alerting procedure before really sending out the e-mail.
# In that timespan, events are collected and will be sent all
# using a single e-mail. Defaults to 10 seconds.
config_properties['MailAlerting.EventCollectTime'] = 10
# Define the minimum time between two alert e-mails in seconds
# to avoid spamming. All events during this timespan are collected
# and sent out with the next report. Defaults to 600 seconds.
config_properties['MailAlerting.MinAlertGap'] = 0
# Define the maximum time between two alert e-mails in seconds.
# When undefined this defaults to "MailAlerting.MinAlertGap".
# Otherwise this will activate an exponential backoff to reduce
# messages during permanent error states by increasing the alert
# gap by 50% when more alert-worthy events were recorded while
# the previous gap time was not yet elapsed.
config_properties['MailAlerting.MaxAlertGap'] = 600
# Define how many events should be included in one alert mail
# at most. This defaults to 1000
config_properties['MailAlerting.MaxEventsPerMessage'] = 1000
# config_properties['LogPrefix'] = 'Original log line: '

# Add your ruleset here:


def build_analysis_pipeline(analysis_context):
    """
    Define the function to create pipeline for parsing the log data.
    It has also to define an AtomizerFactory to instruct py how to process incoming data streams to create log atoms from them.
    """
    # Build the parsing model:
    from aminer.parsing.FirstMatchModelElement import FirstMatchModelElement
    from aminer.parsing.SequenceModelElement import SequenceModelElement
    from aminer.parsing.DateTimeModelElement import DateTimeModelElement
    from aminer.parsing.FixedDataModelElement import FixedDataModelElement
    from aminer.parsing.DelimitedDataModelElement import DelimitedDataModelElement
    from aminer.parsing.AnyByteDataModelElement import AnyByteDataModelElement

    service_children_disk_upgrade = [
        DateTimeModelElement('Date', b'%d.%m.%Y %H:%M:%S'), FixedDataModelElement('UName', b' ubuntu '),
        DelimitedDataModelElement('User', b' '), FixedDataModelElement('HD Repair', b' System rebooted for hard disk upgrade')]

    service_children_home_path = [
        FixedDataModelElement('Pwd', b'The Path of the home directory shown by pwd of the user '),
        DelimitedDataModelElement('Username', b' '), FixedDataModelElement('Is', b' is: '), AnyByteDataModelElement('Path')]

    parsing_model = FirstMatchModelElement('model', [
        SequenceModelElement('Disk Upgrade', service_children_disk_upgrade),
        SequenceModelElement('Home Path', service_children_home_path)])

    # Some generic imports.
    from aminer.analysis import AtomFilters

    # Create all global handler lists here and append the real handlers later on.
    # Use this filter to distribute all atoms to the analysis handlers.
    atom_filter = AtomFilters.SubhandlerFilter(None)

    from aminer.events.StreamPrinterEventHandler import StreamPrinterEventHandler
    stream_printer_event_handler = StreamPrinterEventHandler(None)
    anomaly_event_handlers = [stream_printer_event_handler]

    # Now define the AtomizerFactory using the model. A simple line based one is usually sufficient.
    from aminer.input.SimpleByteStreamLineAtomizerFactory import SimpleByteStreamLineAtomizerFactory
    analysis_context.atomizer_factory = SimpleByteStreamLineAtomizerFactory(
        parsing_model, [atom_filter], anomaly_event_handlers, default_timestamp_path_list=[''], use_real_time=True)

    # Just report all unparsed atoms to the event handlers.
    from aminer.analysis.UnparsedAtomHandlers import SimpleUnparsedAtomHandler
    atom_filter.add_handler(SimpleUnparsedAtomHandler(anomaly_event_handlers), stop_when_handled_flag=True)

    from aminer.analysis.NewMatchPathDetector import NewMatchPathDetector
    new_match_path_detector = NewMatchPathDetector(analysis_context.aminer_config, anomaly_event_handlers, learn_mode=True)
    analysis_context.register_component(new_match_path_detector, component_name=None)
    atom_filter.add_handler(new_match_path_detector)

    from aminer.analysis.NewMatchPathValueComboDetector import NewMatchPathValueComboDetector
    new_match_path_value_combo_detector = NewMatchPathValueComboDetector(analysis_context.aminer_config, [
        '/model/Home Path/Username', '/model/Home Path/Path'], anomaly_event_handlers, learn_mode=True)
    analysis_context.register_component(new_match_path_value_combo_detector, component_name=None)
    atom_filter.add_handler(new_match_path_value_combo_detector)

    # Include the e-mail notification handler only if the configuration parameter was set.
    from aminer.events.DefaultMailNotificationEventHandler import DefaultMailNotificationEventHandler
    if DefaultMailNotificationEventHandler.CONFIG_KEY_MAIL_TARGET_ADDRESS in analysis_context.aminer_config.config_properties:
        mail_notification_handler = DefaultMailNotificationEventHandler(analysis_context)
        analysis_context.register_component(mail_notification_handler, component_name=None)
        anomaly_event_handlers.append(mail_notification_handler)
